apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ApiBaseline.java

```diff
@@ -966,36 +947,37 @@ public void vmChanged(PropertyChangeEvent event) {
 		if (!(event.getSource() instanceof VMStandin)) {
 			String property = event.getProperty();
 			if (IVMInstallChangedListener.PROPERTY_INSTALL_LOCATION.equals(property) || IVMInstallChangedListener.PROPERTY_LIBRARY_LOCATIONS.equals(property)) {
-				try {
-					rebindVM();
-				} catch (CoreException e) {
-					ApiPlugin.log(e);
-				}
+				rebindVM();
 			}
 		}
 	}
 
 	/**
 	 * Re-binds the VM this baseline is bound to.
 	 */
-	private void rebindVM() throws CoreException {
-		fVMBinding = null;
-		IApiComponent[] components = getApiComponents();
-		HashSet<String> ees = new HashSet<>();
-		for (IApiComponent component2 : components) {
-			ees.addAll(Arrays.asList(component2.getExecutionEnvironments()));
-		}
-		resolveSystemLibrary(ees);
+	private void rebindVM() {
+		Job.createSystem("Rebinding JVM", monitor -> { //$NON-NLS-1$
+			try {
+				// Let all the already running job finish first, to avoid errors
+				Job.getJobManager().join(ApiAnalysisJob.class, monitor);
+			} catch (OperationCanceledException | InterruptedException e) {
+				ApiPlugin.log("Interrupted while rebinding JVM", e); //$NON-NLS-1$
+				return;
+			}
+			fVMBinding = null;
+			IApiComponent[] components = getApiComponents();
+			HashSet<String> ees = new HashSet<>();
+			for (IApiComponent component2 : components) {
+				ees.addAll(Arrays.asList(component2.getExecutionEnvironments()));
+			}
+			resolveSystemLibrary(ees);
+		}).schedule();
```

Comment:
should one probably `join()` on the job here? Previously the caller could be sure that the method does not return before the rebind was done.

Reply:
The only caller was a listener. For the listener on JRE changes it is irrelevant to wait on this code to complete.

The reason to use a job here was the possibility of a deadlocks here, after initialize() was made synchronized.